File: lib/imagewriter/d64imagewriter.cc
#include "lib/core/globals.h"
#include "lib/config/flags.h"
#include "lib/data/sector.h"
#include "lib/imagewriter/imagewriter.h"
#include "lib/data/image.h"
#include "lib/external/ldbs.h"
#include "lib/core/logger.h"
#include "lib/config/config.pb.h"
#include <algorithm>
#include <iostream>
#include <fstream>

static int sectors_per_track(int track)
{
    if (track < 17)
        return 21;
    if (track < 24)
        return 19;
    if (track < 30)
        return 18;
    return 17;
}

class D64ImageWriter : public ImageWriter
{
public:
    D64ImageWriter(const ImageWriterProto& config): ImageWriter(config) {}

    void writeImage(const Image& image) override
    {
        log("D64: writing triangular image");

        std::ofstream outputFile(
            _config.filename(), std::ios::out | std::ios::binary);
        if (!outputFile.is_open())
            error("cannot open output file");

        uint32_t offset = 0;
        for (int track = 0; track < 40; track++)
        {
            int sectorCount = sectors_per_track(track);
            for (int sectorId = 0; sectorId < sectorCount; sectorId++)
            {
                const auto& sector = image.get(track, 0, sectorId);
                if (sector)
                {
                    outputFile.seekp(offset);
                    outputFile.write((const char*)sector->data.cbegin(),
                        sector->data.size());
                }

                offset += 256;
            }
        }
    }
};

std::unique_ptr<ImageWriter> ImageWriter::createD64ImageWriter(
    const ImageWriterProto& config)
{
    return std::unique_ptr<ImageWriter>(new D64ImageWriter(config));
}

File: java/com/cowlark/fluxengine/imagewriter/D64ImageWriter.java
package com.cowlark.fluxengine.imagewriter;

import com.cowlark.fluxengine.core.ByteWriter;
import com.cowlark.fluxengine.core.Bytes;
import com.cowlark.fluxengine.data.Image;
import com.cowlark.fluxengine.data.Sector;

/**
 * Writes a D64 (Commodore 1541) sector image, ported from
 * lib/imagewriter/d64imagewriter.cc.
 */
public class D64ImageWriter extends ImageWriter
{
    public D64ImageWriter(ImageWriterProto config)
    {
        super(config);
    }

    private static int sectorsPerTrack(int track)
    {
        if (track < 17)
            return 21;
        if (track < 24)
            return 19;
        if (track < 30)
            return 18;
        return 17;
    }

    @Override
    public void writeImage(Image image)
    {
        System.out.println("D64: writing triangular image");

        Bytes output = new Bytes();
        ByteWriter bw = output.writer();

        int offset = 0;
        for (int track = 0; track < 40; track++)
        {
            int sectorCount = sectorsPerTrack(track);
            for (int sectorId = 0; sectorId < sectorCount; sectorId++)
            {
                Sector sector = image.get(track, 0, sectorId);
                if (sector != null)
                {
                    bw.seek(offset);
                    bw.write(sector.data);
                }

                offset += 256;
            }
        }

        output.writeToFile(config.getFilename());
    }
}
